include/revng/ABIAnalyses/Common.h

```diff
@@ -0,0 +1,213 @@
+  Read,
+  Write,
+  WeakWrite,
+  TheCall,
+  None,
+};
+
+struct ABIAnalysis {
+private:
+  SmallSet<GlobalVariable *, 20> ABIRegisters;
+  SmallVector<GlobalVariable *, 20> RegisterList;
+  const Instruction *CallSite;
+
+public:
+  ABIAnalysis(const GeneratedCodeBasicInfo &GCBI) :
+    ABIAnalysis(nullptr, GCBI){};
+
+  ABIAnalysis(const Instruction *CS, const GeneratedCodeBasicInfo &GCBI) :
+    RegisterList(), CallSite(CS) {
+
+    for (auto *CSV : GCBI.abiRegisters())
+      if (CSV) {
+        ABIRegisters.insert(CSV);
+        RegisterList.emplace_back(CSV);
+      }
+  };
+
+  const SmallVector<GlobalVariable *, 20> getRegisters() const {
+    return RegisterList;
+  }
+
+  bool isABIRegister(const Value *) const;
+
+  TransferKind classifyInstruction(const Instruction *) const;
+
+  SmallVector<const GlobalVariable *, 1>
+  getRegistersWritten(const Instruction *) const;
+
+  SmallVector<const GlobalVariable *, 1>
+  getRegistersRead(const Instruction *) const;
+};
+
+template<typename LatticeElementMap, typename CoreLattice, typename KeyT>
+typename CoreLattice::LatticeElement
+getOrDefault(const LatticeElementMap &S, const KeyT &K) {
+  auto V = S.find(K);
+  if (V != S.end()) {
+    return S.lookup(K);
+  }
+  return CoreLattice::DefaultLatticeElement;
+}
+
+template<typename LatticeElementMap, typename CoreLattice>
+LatticeElementMap
+combineValues(const LatticeElementMap &Lh, const LatticeElementMap &Rh) {
+
+  LatticeElementMap New = Lh;
+  for (const auto &[Reg, S] : Rh) {
+    New[Reg] = CoreLattice::
+      combineValues(getOrDefault<LatticeElementMap, CoreLattice>(New, Reg),
+                    getOrDefault<LatticeElementMap, CoreLattice>(Rh, Reg));
+  }
+  return New;
+}
+
+template<typename LatticeElementMap, typename CoreLattice>
+bool isLessOrEqual(const LatticeElementMap &Lh, const LatticeElementMap &Rh) {
+  for (auto &[Reg, S] : Lh) {
+    if (!CoreLattice::isLessOrEqual(getOrDefault<LatticeElementMap,
+                                                 CoreLattice>(Lh, Reg),
+                                    getOrDefault<LatticeElementMap,
+                                                 CoreLattice>(Rh, Reg))) {
+      return false;
+    }
+  }
+  for (auto &[Reg, S] : Rh) {
+    if (!CoreLattice::isLessOrEqual(getOrDefault<LatticeElementMap,
+                                                 CoreLattice>(Lh, Reg),
+                                    getOrDefault<LatticeElementMap,
+                                                 CoreLattice>(Rh, Reg))) {
+      return false;
+    }
+  }
+  return true;
+}
+
+inline bool ABIAnalysis::isABIRegister(const Value *V) const {
+  if (const auto &G = dyn_cast<GlobalVariable>(V)) {
+    if (ABIRegisters.count(G) != 0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+inline bool isCallSiteBlock(const BasicBlock *B) {
+  if (auto *C = dyn_cast<CallInst>(&*B->getFirstInsertionPt())) {
+    if (C->getCalledFunction()->getName().contains("precall_hook")) {
```

Comment:
Minimize comparisons with names.

I'm a bit confused by the fact that you're exclusively using `precall_hook`.
Aren't you supposed to use `postcall_hook` for forward analyses?

Reply:
so the assumption is documented in `ABIAnalysis.h`

```
/// Run all abi analyses on the outlined function F
/// the outlined function must have all original function calls
/// replaced with a basic block starting with a call to @precall_hook
/// followed by a summary of the side effects of the function
/// followed by a call to @postcall_hook and a basic block terminating
/// instruction
```

So if the block starts with `@precall_hook` it is enough to identify it as a call site.

Forward analyses that start on a call site with start the analysis from the unique successor of a call site block, likewise backwards analyses will start from the unique predecessor, but this is done in the respective analysis cpp file

`isCallSiteBlock` is used to quickly determine if a `write` is a true write or a `weak write`

Reply:
so the assumption is documented in `ABIAnalysis.h`

```
/// Run all abi analyses on the outlined function F
/// the outlined function must have all original function calls
/// replaced with a basic block starting with a call to @precall_hook
/// followed by a summary of the side effects of the function
/// followed by a call to @postcall_hook and a basic block terminating
/// instruction
```

So if the block starts with `@precall_hook` it is enough to identify it as a call site.

Forward analyses that start on a call site with start the analysis from the unique successor of a call site block, likewise backwards analyses will start from the unique predecessor, but this is done in the respective analysis cpp file

`isCallSiteBlock` is used to quickly determine if a `write` is a true write or a `weak write`

Reply:
I'm not very familiar with the IR you're describing.
Could you please paste some relevant parts?

Reply:
Sure, it was introduced very recently to avoid having a special guard value to identify the `TheCall` transfer function.

Turns out this way of splitting the function summaries to a separate block is also useful to identify `weak writes` (otherwise we would need to scan the whole basic block looking for calls to `[pre/post]call_hooks`

from `lib/StackAnalysis/StackAnalysis.cpp:647`
```c++

template<class FunctionOracle>
llvm::BasicBlock *
CFEPAnalyzer<FunctionOracle>::integrateFunctionCallee(llvm::BasicBlock *BB) {
  using namespace llvm;

  // In case the basic block has a call-site, the function call is
  // replaced with i) hooks that facilitate the ABI analysis ii) summary
  // of the registers clobbered by that function. The newly created basic block
  // is returned so that it can be added to the ValueToValueMap.
  BasicBlock *Split = nullptr;
  auto *Term = BB->getTerminator();
  auto *Call = getFunctionCall(Term);

  // If the successor is null, we are dealing with an indirect call.
  auto *Next = getFunctionCallCallee(Term);

  // What are the registers clobbered by the callee?
  const auto &ClobberedRegs = Oracle.getRegistersClobbered(Next);

  // What is the function type of the callee?
  FunctionTypeValue Ty = Oracle.getFunctionType(Next);

  switch (Ty) {
  case FunctionTypeValue::Regular: {
    BranchInst *Br = BranchInst::Create(getFallthrough(Term));
    ReplaceInstWithInst(Term, Br);

    auto *FTy = llvm::FunctionType::get(llvm::Type::getVoidTy(Context), false);
    auto *PreHookOpqF = OFPHooksFunctionCall.get("__precall_pure_function",
                                                 FTy,
                                                 "precall_hook");
    auto *Last = CallInst::Create(PreHookOpqF, "", Br);

    Split = BB->splitBasicBlock(Last->getPrevNode(),
                                BB->getName() + Twine("__summary"));

    // Prevent the store instructions from being optimized out by storing
    // the rval of a call to an opaque function into the clobbered registers.
    OFPRegistersClobbered.addFnAttribute(Attribute::ReadOnly);
    OFPRegistersClobbered.addFnAttribute(Attribute::NoUnwind);

    for (GlobalVariable *Reg : ClobberedRegs) {
      auto *CSVTy = Reg->getType()->getPointerElementType();
      auto *RegsClobbOpqF = OFPRegistersClobbered
                              .get(BB->getParent()->getName(),
                                   CSVTy,
                                   {},
                                   "registers_clobbered");
      new StoreInst(CallInst::Create(RegsClobbOpqF, "", Br), Reg, Br);
    }

    auto *PostHookOpqF = OFPHooksFunctionCall.get("__postcall_pure_function",
                                                  FTy,
                                                  "postcall_hook");
    CallInst::Create(PostHookOpqF, "", Split->getTerminator());
    break;
  }
....
  Call->eraseFromParent();
  return Split;
}

```

Reply:
No, I meant the IR, sorry.

Reply:
So for example this is the IR these analyses are supposed to work on

```llvm
define internal void @root.bb._start__cloned() {
newFuncRoot:
  br label %bb._start__cloned, !dbg !5768

dispatcher.entry.exitStub:                       
  ret void

bb._start__cloned:                                
  call void (%struct.PlainMetaAddress, i64, i32, i8*, ...) 
  .... ; normal block
  br label %bb._start__cloned__summary, !dbg !5796

bb._start__cloned__summary:                       ; <--- special block
  call void @precall_hook.1()                        ; <---- precall hook call
  ; here you would find load / stores normally
  call void @postcall_hook.2()                     ; <---- postcall hook call
  br label %bb._start.0x5__cloned, !dbg !5797

bb._start.0x5__cloned:                            
  call void (%struct.PlainMetaAddress, i64, i32, i8*, ...)
  .... ; normal block
  br label %anypc__cloned, !dbg !5810, !revng.jt.reasons !5811

anypc__cloned:                                    ; preds = %bb._start.0x5__cloned
  br label %dispatcher.entry.exitStub, !dbg !5812, !revng.block.type !5813
}
```